api/providers/tasks.py

```diff
@@ -0,0 +1,592 @@
+        message = 'Bulk upload preparation failure: registration schema [_id={}] not found'.format(schema_id)
+        return handle_internal_error(initiator=initiator, provider=provider, message=message, dry_run=dry_run)
+    except RegistrationSchema.MultipleObjectsReturned:
+        message = 'Bulk upload preparation failure: multiple registration schemas [_id={}] returned'.format(schema_id)
+        return handle_internal_error(initiator=initiator, provider=provider, message=message, dry_run=dry_run)
+
+    # Create the bulk upload job
+    upload = RegistrationBulkUploadJob.create(payload_hash, initiator, provider, schema)
+    logger.info('Creating a registration bulk upload job with [hash={}] ...'.format(upload.payload_hash))
+    if not dry_run:
+        try:
+            upload.save()
+        except ValidationError:
+            sentry.log_exception()
+            message = 'Bulk upload preparation failure: failed to create the job'
+            return handle_internal_error(initiator=initiator, provider=provider, message=message, dry_run=dry_run)
+        upload.reload()
+        logger.info('Bulk upload job created: [pk={}, hash={}]'.format(upload.id, upload.payload_hash))
+    else:
+        logger.info('Dry run: insertion did not happen')
+
+    # Create registration rows for the bulk upload job
+    registration_rows = parsing_output.get('registrations', [])
+    if not registration_rows:
+        message = 'Bulk upload preparation failure: missing registration rows'
+        return handle_internal_error(initiator=initiator, provider=provider, message=message, dry_run=dry_run)
+    initial_row_count = len(registration_rows)
+    logger.info('Preparing [{}] registration rows for bulk creation ...'.format(initial_row_count))
+
+    row_hash_set = set()
+    bulk_upload_rows = []
+    draft_error_list = []
+    try:
+        for registration_row in registration_rows:
+            bulk_upload_row = RegistrationBulkUploadRow.create(
+                upload,
+                registration_row.get('csv_raw', ''),
+                registration_row.get('csv_parsed'),
+            )
+            metadata = bulk_upload_row.csv_parsed.get('metadata', {}) or {}
+            row_external_id = metadata.get('External ID', 'N/A')
+            row_title = metadata.get('Title', 'N/A')
+            # Check duplicates with the database
+            if RegistrationBulkUploadRow.objects.filter(row_hash=bulk_upload_row.row_hash).exists():
+                error = 'Duplicate rows - existing row found in the system'
+                exception = RegistrationBulkCreationRowError(upload.id, 'N/A', row_title, row_external_id, error=error)
+                logger.error(exception.long_message)
+                sentry.log_message(exception.long_message)
+                draft_error_list.append(exception.short_message)
+            # Continue to check duplicates within the CSV
+            if bulk_upload_row.row_hash in row_hash_set:
```

Comment:
I considered `elif` before but decided to do `if` since we don't want to who both database conflicts and CSV duplicates. That's way I have "Continue" here.

e.g. if we have dup rows A1 and A2 in the CSV and A3 in the database, the error message will let user know that
  * A1 and A2 are dup
  * A1 and A2 have conflicts with DB
It saves one extra upload / parsing for this corner case.

Reply:
Gotcha. One alternative to address both concerns might be something like
```python
duped = False
if RegistrationBulkUploadRow.objects.filter(row_hash=bulk_upload_row.row_hash).exists():
    duped = True
    # Other stuff
if bulk_upload_row.row_hash in row_hash_set:
    duped = True
    # ...
if not duped:
    # add/append
```
But this isn't a blocker as long we're not deduping and continuing with the bulk create.